segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/ConcurrencyManager.java

```diff
@@ -0,0 +1,179 @@
+        this.minParallelism = minParallelism;
+        this.maxParallelism = maxParallelism;
+        this.recentTotalWrittenLength = 0;
+        this.recentWriteCount = 0;
+        this.lastSnapshot = new AtomicReference<>();
+        resetSnapshot();
+    }
+
+    //endregion
+
+    //region Operations
+
+    /**
+     * Records the fact that a write with the given length has just completed successfully.
+     *
+     * @param writeLength The length of the write that completed.
+     */
+    void writeCompleted(int writeLength) {
+        synchronized (this) {
+            this.recentWriteCount++;
+            this.recentTotalWrittenLength += writeLength;
+        }
+    }
+
+    /**
+     * Gets a value indicating the current (most recent) degree of parallelism.
+     *
+     * @return The result
+     */
+    int getCurrentParallelism() {
+        return this.lastSnapshot.get().parallelism;
+    }
+
+    /**
+     * Updates the degree of parallelism (if needed), based on information collected since the last time this method was called.
+     * Rules:
+     * * Only changes something if throughput changed significantly since the last time.
+     * * Changes the degree of parallelism in the same direction as throughput.
+     * ** Exception is if both throughput and the fill ratio are declining, in which case nothing is done.
+     *
+     * @return The degree of parallelism, whether it was updated or not.
+     */
+    int updateParallelism() {
+        // Calculate the most recent throughput and Fill Ratio (of those items that have just been written).
+        final long time = this.timeSupplier.get();
+        final Snapshot lastSnapshot = this.lastSnapshot.get();
+        final long elapsedMillis = (time - lastSnapshot.timeStamp) / AbstractTimer.NANOS_TO_MILLIS;
+        if (elapsedMillis < MIN_FREQUENCY_MILLIS) {
+            // No need to do any change.
+            return lastSnapshot.parallelism;
+        } else if (elapsedMillis >= STALE_MILLIS) {
+            // Last snapshot is too old. We can't make any good determination.
+            resetSnapshot();
+            return this.lastSnapshot.get().parallelism;
+        }
+
+        final int recentCount;
+        final long recentLength;
+        synchronized (this) {
+            recentCount = this.recentWriteCount;
+            recentLength = this.recentTotalWrittenLength;
+            this.recentWriteCount = 0;
+            this.recentTotalWrittenLength = 0;
+        }
+
+        final double recentThroughput = (double) recentLength / elapsedMillis;
+        final double recentFillRatio = WriteQueue.calculateFillRatio(recentLength, recentCount);
+        final double throughputDifference = recentThroughput - lastSnapshot.throughput;
+
+        int parallelism = lastSnapshot.parallelism;
+        if (Math.abs(throughputDifference / lastSnapshot.throughput) >= SIGNIFICANT_DIFFERENCE) {
```

Comment:
Educated guess. I don't want changes due to noise, and I do not want to make it too high, in which case it would never change.

It jumps by 30-50% when the concurrency is far from the optimal one, but as it gets closer, it changes by a smaller margin, and it plateaus at one point. When it plateaus it usually stays within a +-10% range, during which I don't want it to change.  That is the base for my educated guess.